crates/ruma-appservice-api/src/event/push_events.rs

```diff
@@ -2,11 +2,213 @@
 //!
 //! Endpoint to push an event (or batch of events) to the application service.
 
+#[cfg(feature = "unstable-msc2409")]
+use js_int::UInt;
+#[cfg(feature = "unstable-msc2409")]
+use ruma_common::{
+    events::receipt::Receipt, presence::PresenceState, serde::from_raw_json_value, OwnedEventId,
+    OwnedRoomId, OwnedUserId,
+};
+#[cfg(feature = "unstable-msc2409")]
+use serde::{de, Deserialize, Serialize};
+#[cfg(feature = "unstable-msc2409")]
+use serde_json::{value::RawValue as RawJsonValue, Value as JsonValue};
+#[cfg(feature = "unstable-msc2409")]
+use std::collections::BTreeMap;
+
+/// Type for passing ephemeral data to homeservers.
+#[derive(Clone, Debug, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+#[serde(tag = "type", content = "content")]
+pub enum Edu {
+    /// An EDU representing presence updates for users of the sending homeserver.
+    #[serde(rename = "m.presence")]
+    Presence(PresenceContent),
+
+    /// An EDU representing receipt updates for users of the sending homeserver.
+    #[serde(rename = "m.receipt")]
+    Receipt(ReceiptContent),
+
+    /// A typing notification EDU for a user in a room.
+    #[serde(rename = "m.typing")]
+    Typing(TypingContent),
+
+    #[doc(hidden)]
+    _Custom(JsonValue),
```

Comment:
Did you copy this pattern of `serde(tag)` + `_Custom(JsonValue)` from somewhere else? It has some unfortunate consequences on the behavior of serde roundtrips, so I think this variant should also have the `#[serde(skip)]` attribute, or the type shouldn't implement `Serialize` at all (I think I'd prefer the latter).

We could even consider not having it at all and use `Box<RawJsonValue>` for EDUs, like we do for PDUs because they are really hard to represent and Conduit does its own thing anyways.

Wdyt?

Reply:
Yeah, ruma's serde magic is well beyond what I'm used to, so I've taken a lot of inspiration from prior art.

I'm not really familiar enough with Ruma's wider design philosophy to have a preference on this, but if we already use `Box<RawJsonValue>` for PDUs, I suppose it would make sense to go with that approach for EDUs for consistencys sake?

Reply:
Well part of the reason we use `Box<RawJsonValue>` for PDUs (for now) is that they are just very hard to model. This doesn't seem to be as bad for EDUs.

Reply:
Fair enough, I went with not implementing `Serialize`.